url270fc4f5ee_ykcilborw_Joust_tgz-pJ8-fri_patterns_interpreter_parsergenerator_examples_HowToOverrideLexerImplJ8/src/fri/patterns/interpreter/parsergenerator/syntax/builder/SyntaxSeparation.java

```diff
@@ -0,0 +1,339 @@
+package fri.patterns.interpreter.parsergenerator.syntax.builder;
+
+import java.util.*;
+import fri.patterns.interpreter.parsergenerator.Token;
+import fri.patterns.interpreter.parsergenerator.syntax.*;
+import fri.patterns.interpreter.parsergenerator.lexer.StandardLexerRules;
+
+/**
+	Separates lexer from parser rules. Provides token and ignored symbols
+	after construction, which can be used for LexerBuilder construction.
+	<p>
+	Example:
+	<pre>
+		SyntaxSeparation separation = new SyntaxSeparation(new Syntax(myRules));
+		LexerBuilder builder = new LexerBuilder(separation.getLexerSyntax(), separation.getIgnoredSymbols());
+		Lexer lexer = builder.getLexer();
+		// when using the lexer standalone, you must put the token terminal symbols into it now:
+		lexer.setTerminals(separation.getTokenSymbols());
+		// when using a Parser the Parser will do this:
+	</pre>
+	
+	@see fri.patterns.interpreter.parsergenerator.lexer.LexerBuilder
+	@author (c) 2002, Fritz Ritzberger
+*/
+
+public class SyntaxSeparation
+{
+	private List tokenSymbols;
+	private List ignoredSymbols;
+	private Syntax parserSyntax;
+	private Syntax lexerSyntax;
+	public static boolean DEBUG = true;
+
+	/** Splits a syntax into two syntaxes containing parser and lexer rules, tokens and ignored get collected. */
+	public SyntaxSeparation(Syntax syntax)
+		throws SyntaxException
+	{
+		separate(syntax, new IntArray(syntax.size()));
+	}
+
+	/** Returns the lexer syntax that remains after removing all parser rules and token/ignored directives. */
+	public Syntax getLexerSyntax()	{
+		return lexerSyntax;
+	}
+
+	/** Returns the parser syntax that remains after removing all lexer rules. All lexer tokens are marked with `backquotes`. */
+	public Syntax getParserSyntax()	{
+		return parserSyntax;
+	}
+
+	/**
+		Returns the top level token symbols for the lexer. These symbols ARE enclosed within `backquotes`.
+		This can be used at <i>setTerminals</i> call for a standalone Lexer.
+	*/
+	public List getTokenSymbols()	{
+		return tokenSymbols;
+	}
+
+	/**
+		Returns the top level ignored token symbols for the lexer. These symbols are NOT enclosed within `backquotes`.
+		This will be used by the LexerBuilder for feeding the LexerImpl with ignored symbols.
+	*/
+	public List getIgnoredSymbols()	{
+		return ignoredSymbols;
+	}
+
+
+	private void separate(Syntax syntax, IntArray deleteIndexes)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 26. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 76-101 into a separate method.